management-api-agent-common/src/main/java/com/datastax/mgmtapi/NodeOpsProvider.java

```diff
@@ -740,39 +737,46 @@ public void clearSnapshots(
   @Rpc(name = "repair")
   public String repair(
       @RpcParam(name = "keyspaceName") String keyspace,
-      @RpcParam(name = "tables") Optional<List<String>> tables,
+      @RpcParam(name = "tables") List<String> tables,
       @RpcParam(name = "full") Boolean full,
       @RpcParam(name = "notifications") boolean notifications,
-      @RpcParam(name = "repairParallelism") Optional<RepairParallelism> repairParallelism,
-      @RpcParam(name = "datacenters") Optional<Collection<String>> datacenters,
-      @RpcParam(name = "associatedTokens") Optional<List<RingRange>> associatedTokens,
-      @RpcParam(name = "repairThreadCount") Optional<Integer> repairThreadCount)
+      @RpcParam(name = "repairParallelism") String repairParallelism,
```

Comment:
Question: Is it not possible to use an enum here? We could probably source the enum from Cassandra to avoid direct dependancies between the agent and server.

Reply:
> Question: Is it not possible to use an enum here? We could probably source the enum from Cassandra to avoid direct dependencies between the agent and server.

It is possible, however it would have to be defined in the `management-api-agent-common` or `management-api-common` projects and then also referenced in the `management-api-server` project. It might also have to be added to all of the flavors of `GenericSerializerXYZ` in order for it to be serialized/deserialized correctly.

Using 3.11 to demonstrate, if you look at the RpcMethod stuff here:
https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-agent-3.x/src/main/java/com/datastax/mgmtapi/rpc/RpcMethod3x.java#L79-L82
you will see where the serializers for any RPC method arguments are setup.

A little more complicated looking, but the same basic idea for serializers for returned objects from RPC methods is here:
https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-agent-3.x/src/main/java/com/datastax/mgmtapi/rpc/RpcMethod3x.java#L91-L104

Basically, RPC method arguments will be handled by GenericSerializer, and RPC methods return values are handled by ObjectSerializer.

Regarding RPC method arguments, we can look at the 3.11 GenericSerializer code:
https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-agent-3.x/src/main/java/com/datastax/mgmtapi/rpc/GenericSerializer3x.java#L33-L67
and
https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/master/management-api-agent-3.x/src/main/java/com/datastax/mgmtapi/rpc/GenericSerializer3x.java#L102-L142
You'll see that it basically has to map Cassandra types to/from Java types, and it also handles Java Lists, Sets and Maps. If we want to handle an enum, it is going to have to be added in here so that the argument can be serialized/deserialzed correctly for the RPC call. This is the slippery slope I mentioned in the other PR. We could keep adding and adding and adding and it will get a bit unruly. Looking at the code, it seems this was intended to handle basic types, with the small addition of Lists, Sets and Maps.

You might have the desire to then use the Cassandra type `RepairParallelism` here, since it is an enum and already defined in Cassandra. There is a big potential problem with that though. The problem is that anything defined in Cassandra potentially needs to be compiled against Cassandra specific jarfiles. This is why we have so many agent projects. The code is mostly the same (not always), but if it's compiled against the wrong version of Cassandra jarfiles, the agents produced fail when deployed in Cassandra instances of different versions than what it was compiled against. We currently do NOT have a dependency upon Cassandra jarfiles in the management-api-server project, and we need to keep it that way. If we introduce Cassandra dependnecies there, then we will have just as many server projects as we have agent projects.

Reply:
OK, got it. Sounds like a string is going to be the most efficient way to do this 😫